### PR TITLE
Add strip controller support

### PR DESCRIPTION
**Pull request by autinerd**
This pull request will add support for Magic Home RGB strip controllers.

### CODE REVIEW

**Comment by alexvandervegt** [comment]
When will this be merged? My light is broken a long time now...

**Comment by Danielhiversen** [comment]
I will look at the code when the tests are passing

**Comment by alexvandervegt** [comment]
Thanks, hopefully this will be soon :)

**Comment by alexvandervegt** [comment]
@autinerd?

**Comment by alexvandervegt** [comment]
@autinerd do you have a time frame for us?

**Comment by Danielhiversen** [comment]
Please stop spamming the pr.

**Comment by sehraf** [comment]
*gentle ping*
I've run the test script and while it detected everything correct, the effect (7 colors quick) wasn't quick (was more the slow version of it)
`[[0, 1], [1, 1], [2, 1], [3, 0], [5, 1], [7, 1], [9, 1], [10, 0]]`
Beside this it was able to control the strip.

File: flux_led/__main__.py [hunk 1]
```diff
@@ -711,21 +788,32 @@ def update_state(self, retry=2 ):
...
+            mode = self._determineMode(ww_level, pattern)
+            if mode == "unknown":
+                if retry < 1:
+                    return
+                self.update_state(max(retry-1, 0))
                 return
-            self.update_state(max(retry-1, 0))
-            return
+        else:
+            pattern = rx[3] << 8 + rx[4]
```
Comment (sehraf) [inline flux_led/__main__.py]: This is missing an bracket around the shift.

**Comment by sehraf** [comment]
I had to apply some minor patch to get a working version:
```diff
diff --git a/flux_led/__main__.py b/flux_led/__main__.py
index 31a67cc..3cc6b7e 100755
--- a/flux_led/__main__.py
+++ b/flux_led/__main__.py
@@ -630,6 +630,8 @@ class WifiLedBulb():
             mode = "color"
         elif PresetPattern.valid(pattern_code):
             mode = "preset"
+        elif PresetPattern.valid_strip(pattern_code):
+            mode = "custom"
         elif BuiltInTimer.valid(pattern_code):
             mode = BuiltInTimer.valtostr(pattern_code)
         return mode
@@ -799,7 +801,7 @@ class WifiLedBulb():
                 self.update_state(max(retry-1, 0))
                 return
         else:
-            pattern = rx[3] << 8 + rx[4]
+            pattern = (rx[3] << 8) + rx[4]
             mode = self._determineMode(0, pattern)
             if mode == "unknown":
                 if retry < 1:

```
The check for the mode needs some love, at least it's not crashing anymore.

**Comment by sehraf** [comment]
I did some tinkering and testing and found a regression: `_send_msg` is appending to the parameter `bytes` and thus modifying caller's data!
This resulted in weird side effects, like the strip turned white and HA took ages to update the state. 
With this small change (anybody up for using `copy()`?) HA integration works fine (color and state wise, effects are right now out of scope but i'm working on it)

```diff
@@ -1090,12 +1106,16 @@ class WifiLedBulb():
         return colorsys.hsv_to_rgb(hsv[0], hsv[1], level)
 
     def _send_msg(self, bytes, checksum=True):
+        # create a copy! Otherwise we append to bytes and modify the callers data, too.
+        b2 = bytearray()
+        b2[:] = bytes
+
         # calculate checksum of byte array and add to end
         if checksum and self._use_csum:
             csum = sum(bytes) & 0xFF
-            bytes.append(csum)
+            b2.append(csum)
         with self._lock:
-            self._socket.send(bytes)
+            self._socket.send(b2)
 
     def _read_msg(self, expected):
         remaining = expected
```

**Comment by vquelque** [comment]
This is exactly the code I was looking for ! I was wondering how to trigger these preset ! Glad I found this PR !! I don't know why this still not have been merged. The code with @sehraf edits is working great for me except preset speed which is not working accordingly (everything is very slow). I'm going to investigate this issue.

**Comment by sehraf** [comment]
nice to hear that it is useful to somebody :+1: 
I'm struggling with some weird behavior when controlling the strip with home assistant (it turns white at a maybe 60% brightness). 
Beside fixing this, effects are on my todo, too. (Also, if anybody knows how to get all 300 names of the effect, let me know)

**Comment by lstg** [comment]
@sehraf looks like you're making some good progress :) See https://github.com/home-assistant/core/issues/22161#issuecomment-478550467 for the full list of effects

**Comment by sehraf** [comment]
That was exactly what i was looking for! I can also confirm that speed readings are broken...
Presets are now correctly named, i'll prepare a git branch instead of posting patches here.
```
ON [Pattern: Seven_colors_change_quickly (Speed 0%) raw state: b'81a1230067330000ff0303003115' raw strip state: b'63003104000000000000039b']
ON [Pattern: Seven_colors_change_quickly (Speed 0%) raw state: b'81a1230067330000ff0303003115' raw strip state: b'63003104000000000000039b']
ON [Pattern: Seven_colors_strobe_flash (Speed 0%) raw state: b'81a1230068330000000303003117' raw strip state: b'63003104000000000000039b']
ON [Pattern: Seven_colors_strobe_flash (Speed 0%) raw state: b'81a1230068330000000303003117' raw strip state: b'63003104000000000000039b']
```

**Comment by vquelque** [comment]
I was quite busy since last week so I didn't get back. I managed to fix the issue with the speed. It is working great on my side with the controller I have (magic home LED SPI controller). The difference with the original code is that my controller directly sends/receives an hex value ranging from 0x01 to 0x64 (1% to 100%) in byte 5 for controlling the speed instead of this weird "delay" thing. I patched the code and it is working great so far. I thing I still need to clean the code a little bit maybe and maybe optimize some things when I have time. You can find the patched file in my repo. It would be really nice if you do a PR ! I can help too :)
Thanks a lot @sehraf for the fixes you already did !

**Comment by vquelque** [comment]
Ant thank you @lstg for providing the effect list ! I was going to dump the strings from the apk but it's easier now.

**Comment by sehraf** [comment]
@vquelque  Sure thing! I have a look and see how my hackish code can be integrated.
I have the same controller btw :)

**Comment by vquelque** [comment]
@sehraf at least we have the same fix haha 🤣

**Comment by sehraf** [comment]
@vquelque there you go https://github.com/vquelque/flux_led/pull/1

I feel stupid that I've programmed the speed thing instead of just plain `int()` :roll_eyes: 

For those who are interested: In https://github.com/sehraf/flux_led/tree/WIP you can find my hacked-together code. It also includes a (started) reimplementation using classes instead of `if` jungle to distinguish the different versions. The goal is to get a stable home assistant integration....

**Pull request closed without merging**